### ISSUE TITLE
Code doesn't recognize "game solved" in rare cases

### ISSUE THREAD

**Issue by pointsman** [calancha/Minesweeper#2]
The game doesn't recognize in all cases, that the game is solved.

Since this happend to me already in a few cases in the past, I added
some logging to the (otherwise unmodified) code. This is the log of a
game in default configuration (8 rows, 8 columns, 10 bombs) for which
mines.el didn't recognized that the game is solved.

    Start: (17 21 26 38 40 46 55 56 58 62)
    Uncover: 56
    Avoided game over in the first move
    State after avoided game over in the first move: (17 21 26 38 40 46 55 0 58 62)
    Uncover: 63
    Uncover: 7
    Uncover: 29
    Uncover: 28
    Uncover: 27
    Uncover: 35
    Uncover: 36
    Uncover: 39
    Uncover: 47
    Uncover: 54
    Uncover: 32
    Uncover: 24
    Uncover: 25
    Uncover: 16
    Uncover: 8

**Comment by calancha** [comment]
Thank you very much!
I suffered this few times as well.
The reason seems to be **mines-end-p** function.  The line
```
(equal mines-mine-positions (mines--find-pos nil mines-state))
```

is asking for problems...  I mean, that is a sufficient condition to end
the game but it's not a necessary one.

A necessary and sufficient condition to end the game can be proved using the
*Dirichlet's box principle* (a.k.a *Pigeonhole principle*).
* We have n boxes, m of them containing a bomb (n > m).
* We have already opened n - m boxes and we are still alive.

Then, for the *Dirichlet principle* we prove that the m bombs are
in the remaining m boxes.

The `equal' comparison above requires somehow an specific order to open
the m boxes: the fact is that the bombs are there regardless of the order
you chose to open them.

Therefore, the fix I have in mind is to end the game whenever
**mines-number-mines** equals to the remaining uncovered cells.  That is,
implement **mines-end-p** as follows:
 ```
(defun mines-end-p ()
  "Return non-nil when the game is completed.
The game is completed once the only uncovered cells are those
having a mine."
  (= mines-number-mines (length (mines--find-pos nil mines-state))))
```

I have pushed a new branch with the fix:
https://github.com/calancha/Minesweeper/tree/detect-end-game-fails

If we don't see the problem again in this new branch in a while (~ 1 week),
then I will release version 1.6 with this fix.
